### ISSUE TITLE
Has hdfs connector done something to handle small files

### ISSUE THREAD

**Issue by jinlaimm** [confluentinc/kafka-connect-hdfs#271]
Hi,
When reading the doc, i see that there are a few parameters to decide the frequency of writing hdfs (flush.size, rotate.interval.ms, rotate.schedule.interval.ms and so on).
I'm wondering if there has some parameters to handle with small files on hdfs.

Thanks.
Mengmeng

**Comment by cricket007** [comment]
You control file sizes by increasing the flush and/or rotation. 

You can find various Spark or MapReduce processes such as LinkedIn's Camus Sweeper or Gobblin, which can efficiently roll up the files separately from Kafka Connect by reading previously written files.

**Comment by alexnu** [comment]
So this means that it's best to handle the hadoop small files problem outside Kafka Connect, right? For example schedule a job to merge small files into bigger ones.

**Comment by cricket007** [comment]
Correct. It also would make sense to have the same job de-dupe data. For
example, if you're sinking a compacted topic.

FWIW, I've been able to get 4 GB Avro  files written with Connect by
setting a large flush size & rotation interval. So, if you don't expect the
data in real time, then you can do the same.

**Comment by cricket007** [comment]
https://github.com/KeithSSmith/spark-compaction  
https://github.com/edwardcapriolo/filecrush  

[Camus](https://docs.confluent.io/current/camus/docs/intro.html) was the solution before HDFS Connect, and superseded by Gobblin. it had it's own compaction processes
https://github.com/allegro/camus-compressor  
https://github.com/confluentinc/camus/tree/confluent-master/camus-sweeper

http://gobblin.readthedocs.io/en/latest/user-guide/Compaction/

**Comment by alexnu** [comment]
Wow thanks for the info, this was really helpful. 

I was considering using [hadoop streaming](http://archive-primary.cloudera.com/cdh5/cdh/5/hadoop-2.6.0-cdh5.9.1/hadoop-mapreduce-client/hadoop-mapreduce-client-core/HadoopStreaming.html) for compaction, but you provided us with a lot more alternatives. Plus we are already familiar with Gobblin, as we are currently using it to sink from Kafka.

**Comment by cricket007** [comment]
Welcome. 

Not sure how Hadoop Streaming would help. The files would already be on HDFS, and you aren't piping them into stdin. At that point, you are just running MapReduce.

**Comment by alexnu** [comment]
Well yes, but if run with only one reducer it could `cat` all small files into a single large file.

Isn't that what the tools you suggested do ?
https://github.com/KeithSSmith/spark-compaction
https://github.com/edwardcapriolo/filecrush

**Comment by alexnu** [comment]
I should probably have mentioned that I'm working with text files, sorry about that.

**Comment by cricket007** [comment]
Right, it's a step after outside the scope of Kafka Connect. 

Another alternative would be a NiFi Flow of 

`GetHDFS => MergeContent => (CompressContent =>) PutHDFS`
